drivers/docker/reconciler.go

```diff
@@ -0,0 +1,174 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+func (d *Driver) removeDanglingContainersGoroutine() {
+	if !d.config.GC.DanglingContainers.Enabled {
+		d.logger.Debug("skipping dangling containers handling; is disabled")
+		return
+	}
+
+	period := d.config.GC.DanglingContainers.period
+
+	succeeded := true
+
+	// ensure that we wait for at least a period or creation timeout
+	// for first container GC iteration
+	// The initial period is a grace period for restore allocation
+	// before a driver may kill containers launched by an earlier nomad
+	// process.
+	initialDelay := period
+	if d.config.GC.DanglingContainers.creationTimeout > initialDelay {
+		initialDelay = d.config.GC.DanglingContainers.creationTimeout
+	}
+
+	timer := time.NewTimer(initialDelay)
+	for {
+		select {
+		case <-timer.C:
+			if d.previouslyDetected() && d.fingerprintSuccessful() {
+				err := d.removeDanglingContainersIteration()
+				if err != nil && succeeded {
```

Comment:
I'm not totally sure I understand this `succeeded` flag. The `removeDanglingContainersIteration` logs quite a bit but only returns a wrapped error if the `client.ListContainers` returns an error. Is this just to reduce log noise?

Reply:
Yes, just to reduce log noise.  `removeDanglingContainersIteration` logs for specific containers only for debugging purposes.  Here, I want to protect against the super noisy case where docker daemon is unexpectedly hang and somehow end up logging a failure log line every period.

Given that we check for successful fingerprinting just seconds before, I can understand that this is potentially unnecessary complexity - what do you think?

Reply:
I've seen that sort of thing happen myself, so I agree it's something to consider. But not dropping log lines seems like it could potentially mislead the operator. At the risk of adding complexity: maybe the right thing to do here is to back off on attempting to remove the containers, rather than backing off on the logging?

Reply:
I'm using both strategies here.  First backing off removal of containers by checking `fingerprintSuccessful` first; and then backoff logging if it constantly fails.

We've had some user complaints about logging for periodic maintenance operations, e.g. logging fingerprints before and triggering log churn/retention/cost issues; so I got motivated to reduce logging in this periodic operation here.  Considering that we still log failures on their first occurrence, I'm not sure how misleading it is.

Though, I don't have a good way to balance concern here, so I could be persuaded to switch.

Reply:
Given that the default period for this loop is 5 minutes, I think you're probably right and I'm overthinking this. 😀